### PR TITLE
[Analysis] Add cost model for experimental.cttz.elts intrinsic

### PR DESCRIPTION
**Pull request by david-arm**
In PR #88385 I've added support for auto-vectorisation of some early exit loops, which requires using the experimental.cttz.elts to calculate final indices in the early exit block. We need a more accurate cost model for this intrinsic to better reflect the cost of work required in the early exit block. I've tried to accurately represent the expansion code for the intrinsic when the target does not have efficient lowering for it. It's quite tricky to model because you need to first figure out what types will actually be used in the expansion. The type used can have a significant effect on the cost if you end up using illegal vector types.

Tests added here:

  Analysis/CostModel/AArch64/cttz_elts.ll
  Analysis/CostModel/RISCV/cttz_elts.ll

### CODE REVIEW

**Comment by sjoerdmeijer** [comment]
I haven't looked or seen much codegen yet for this intrinsic, but is this "as cheap as an add instruction"?

**Comment by david-arm** [comment]
> I haven't looked or seen much codegen yet for this intrinsic, but is this "as cheap as an add instruction"?

Definitely not, but then I've not written it to be as cheap as an add instruction either. If we have to expand the intrinsic then the cost will be quite high. If we're not expanding the intrinsic then the only sensible thing that the BasicTTIImpl implementation can do is assume it's cheap and it's then up to the target cost model to override this in it's own getIntrinsicIntrCost function.

**Comment by sjoerdmeijer** [comment]
> > I haven't looked or seen much codegen yet for this intrinsic, but is this "as cheap as an add instruction"?
> 
> Definitely not, but then I've not written it to be as cheap as an add instruction either. If we have to expand the intrinsic then the cost will be quite high. If we're not expanding the intrinsic then the only sensible thing that the BasicTTIImpl implementation can do is assume it's cheap and it's then up to the target cost model to override this in it's own getIntrinsicIntrCost function.

Ah, sorry, I didn't look properly, I didn't expand `llvm/test/Analysis/CostModel/AArch64/cttz_elts.ll` and only saw an estimated costs of 1 but that's the other test case. It all makes sense now. I will go over the rest of the changes now.

**Review by sjoerdmeijer — APPROVED** [review]
The AArch64 changes look reasonable, looks like an improvement, so LGTM.
Maybe wait a couple of days in case @topperc has comments about the RISC-V changes.

File: llvm/test/Analysis/CostModel/RISCV/cttz_elts.ll [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %res.i32.nxv8i1.nzip = call i32 @llvm.experimental.cttz.elts.i32.nxv8i1(<vscale x 8 x i1> undef, i1 false)
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %res.i32.nxv16i1.nzip = call i32 @llvm.experimental.cttz.elts.i32.nxv16i1(<vscale x 16 x i1> undef, i1 false)
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %res.i32.nxv32i1.nzip = call i32 @llvm.experimental.cttz.elts.i32.nxv32i1(<vscale x 32 x i1> undef, i1 false)
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: ret void
+;
+  %res.i64.nxv2i1.zip = call i64 @llvm.experimental.cttz.elts.i64.nxv2i1(<vscale x 2 x i1> undef, i1 true)
+  %res.i64.nxv4i1.zip = call i64 @llvm.experimental.cttz.elts.i64.nxv4i1(<vscale x 4 x i1> undef, i1 true)
+  %res.i64.nxv8i1.zip = call i64 @llvm.experimental.cttz.elts.i64.nxv8i1(<vscale x 8 x i1> undef, i1 true)
+  %res.i64.nxv16i1.zip = call i64 @llvm.experimental.cttz.elts.i64.nxv161(<vscale x 16 x i1> undef, i1 true)
+  %res.i64.nxv32i1.zip = call i64 @llvm.experimental.cttz.elts.i64.nxv32i1(<vscale x 32 x i1> undef, i1 true)
```
Comment (topperc) [inline llvm/test/Analysis/CostModel/RISCV/cttz_elts.ll:32]: Can we test <vscale x 64 x i1> and <vscale x 128 x i1>? <vscale x 128 x i1> is not a legal type so will require splitting

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**